### PR TITLE
[const prop] Fix "alloc id without corresponding allocation" ICE

### PR DESCRIPTION
**Pull request by wesleywiser**
r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/interpret/intern.rs [hunk 1]
```diff
@@ -265,12 +312,25 @@ for
...
-    ecx: &mut CompileTimeEvalContext<'mir, 'tcx>,
+pub fn intern_const_alloc_recursive<M>(
+    ecx: &mut InterpCx<'mir, 'tcx, M>,
     // The `mutability` of the place, ignoring the type.
     place_mut: Option<hir::Mutability>,
     ret: MPlaceTy<'tcx>,
-) -> InterpResult<'tcx> {
+) -> InterpResult<'tcx>
+where
+    M: Machine<
```
Comment (Centril) [inline src/librustc_mir/interpret/intern.rs]: A trait alias would be great for this and the uses above. ;)

File: src/librustc_mir/transform/const_prop.rs [hunk 2]
```diff
@@ -649,20 +649,31 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
...
+        let is_scalar = match *op {
             interpret::Operand::Immediate(Immediate::Scalar(ScalarMaybeUndef::Scalar(s))) =>
                 s.is_bits(),
             interpret::Operand::Immediate(Immediate::ScalarPair(ScalarMaybeUndef::Scalar(l),
                                                                 ScalarMaybeUndef::Scalar(r))) =>
                 l.is_bits() && r.is_bits(),
             _ => false
+        };
+
+        if let interpret::Operand::Indirect(_) = *op {
```
Comment (oli-obk) [inline src/librustc_mir/transform/const_prop.rs]: This could be an arm in the above match

**Comment by oli-obk** [comment]
Do we have any tests actually exercising the creation of a new const alloc and const propping a pointer to it without triggering promotion? I haven't looked into it yet, but I now also worry about interior mutability behind references.

**Comment by wesleywiser** [comment]
> Do we have any tests actually exercising the creation of a new const alloc and const propping a pointer to it without triggering promotion?

I tried a couple different things and couldn't get that to happen. I think the blocker is this bit of code

https://github.com/rust-lang/rust/blob/388ffd9df83cfac20b3fbf7128eb44d5b1526f8c/src/librustc_mir/transform/const_prop.rs#L618

**Pull request merged**